### ISSUE TITLE
Allow configuring UnnecessaryFullyQualifiedName to ignore certain packages

### ISSUE THREAD

**Issue by eygraber** [detekt/detekt#9279]
Currently, `UnnecessaryFullyQualifiedName` reports a violation for `android.R.id.content`, which is correct, because the class name is `R`, and the FQN is `android.R`.

Since that pattern (and others) is a common occurrence, it would be nice if we could specify a list of packages that would be excluded from the rule.

**Comment by marpalat** [comment]
Hi! I'm new to open source contributions and not yet familiar with this repo's rules, but I would be glad to implement this feature, if that's OK.

**Comment by BraisGabin** [comment]
Sure, go for it!

**Comment by BraisGabin** [comment]
I'm thinking about this... Should we ignore packages or classes?

For example, for your use case we should allow `android.R`. Maybe allow everything inside `android` package is too much.

I'm thinking about how the configuration API should be.  This is an idea:

```yaml
excludeClasses:
  - android.R # only this class
  - android.* # every class in the android package. No subpackages
  - android.**.* # every class in a subpackage of android
```

**Comment by marpalat** [comment]
That could be a good addition to the original proposal. I’m not sure, how decisions are made in open source. Let’s see, what others say, right?

**Comment by marpalat** [comment]
Maybe a more granular config would be a good match in this case? Something like: `ignoreMembers`, `ignoreClasses`, `ignorePackages`. More specific filters are applied first. This gives quite powerful and flexible options, and also from the user perspective it could be more clear.

**Comment by eygraber** [comment]
> For example, for your use case we should allow android.R. Maybe allow everything inside android package is too much.

That's a good point

**Comment by BraisGabin** [comment]
About the name: `excludeClasses` is a bad idea because there are also top level functions or properties. And I would only add one configuration. What do you think about `ignoredFullyQualifiedNames`?

**Comment by marpalat** [comment]
>  What do you think about ignoredFullyQualifiedNames?

 This name is much clearer in my understanding, but as a beginner in Detekt, I'm not sure, how easy it will be to parse the config in such a way that we can check all those different things (like members, classes, top level functions) against this config. It definitely will need a bit of investigation (and time) from my side.

**Comment by BraisGabin** [comment]
We have glob support on other rules so that shouldn't be difficult to implement. You can use the extension function `String.simplePatternToRegex()`. And you can find the usages of it to get an idea about how to use it.

**Comment by marpalat** [comment]
It looks like` String.simplePatternToRegex()` is not the right choice in this situation (or it works differently, than expected). It doesn't take segments into account.
E.g.` java.nio.charset.StandardCharsets` is matched by pattern `"java.nio.*"`, but according to provided requirements it shouldn't, because` "java.nio.*"` should not match classes in sub-packages.

**Comment by marpalat** [comment]
So, I'm not sure, what I should do here.

1. Either I can use `String.simplePatternToRegex()` similarly to other rules in the repo, and in this case we will have 2 configuration options:
```
ignoredFullyQualifiedNames:
  - android.R # only this class
  - android.* # everything in the android package including sub-packages
  ``` 

2. Or I can write a custom matcher, that will allow more detailed configuration for sub-packages as described above:
```
ignoredFullyQualifiedNames:
  - android.R # only this class
  - android.* # everything in the android package. No subpackages
  - android.**.* # everything in a subpackage of android
  ```

**Comment by BraisGabin** [comment]
There is an implementation of `qualifiedNameGlobToRegex` in `AnnotationSuppressor.kt`. I'm going to move it out of there because, as we can see here, that code is useful somewhere else. That should unlock this.

**Comment by BraisGabin** [comment]
@marpalat #9340 should unblock you. Once it is merged you would be able to use `fullyQualifiedNameGlobToRegex`.

**Comment by BraisGabin** [comment]
Merged :)

**Comment by cortinico** [comment]
> Merged :)

@BraisGabin can we close this issue then?

**Comment by BraisGabin** [comment]
No, i just merged a pr that was blocking this one.

**Comment by marpalat** [comment]
Good 😎 Thank you @BraisGabin ! I’ll rebase and apply the newly moved function, once I have some free time.

**Comment by Phoenix1808** [comment]
is this issue open or sorted ?

**Comment by marpalat** [comment]
I’m waiting for a PR review now.

**Comment by BraisGabin** [comment]
This was fixed with #9301
